src/ZkFold/Symbolic/Data/ByteString.hs

```diff
@@ -75,8 +206,122 @@ instance (KnownNat p, KnownNat n) => BoolType (ByteString n (Zp p)) where
     -- | Bitwise and
     x && y = fromConstant @Natural $ toConstant x .&. toConstant y
 
+    -- | Bitwise xor
+    xor x y = fromConstant @Natural $ toConstant x `B.xor` toConstant y
+
+
+-- | A ByteString of length @n@ can only be split into words of length @wordSize@ if all of the following conditions are met:
+-- 1. @wordSize@ is not greater than @n@;
+-- 2. @wordSize@ is not zero;
+-- 3. The bytestring is not empty;
+-- 4. @wordSize@ divides @n@.
+--
+instance
+  ( KnownNat wordSize
+  , KnownNat n
+  , KnownNat p
+  , wordSize <= n
+  , 1 <= wordSize
+  , 1 <= n
+  , Mod n wordSize ~ 0
+  ) => ToWords (ByteString n (Zp p)) (ByteString wordSize (Zp p)) where
+
+    toWords bs = fmap fromConstant $ reverse $ take (Haskell.fromIntegral $ n `Haskell.div` wordSize) natWords
+      where
+        asNat :: Natural
+        asNat = toConstant bs
+
+        n :: Natural
+        n = getNatural @n
+
+        wordSize :: Natural
+        wordSize = getNatural @wordSize
+
+        natWords :: [Natural]
+        natWords = unfoldr (toBase (2 Haskell.^ wordSize)) asNat <> Haskell.repeat (fromConstant @Natural 0)
+
+-- | Unfortunately, Haskell does not support dependent types yet,
+-- so we have no possibility to infer the exact type of the result
+-- (the list can contain an arbitrary number of words).
```

Comment:
Maybe, a `Vector k a` can be used instead of `[a]`, then?

Reply:
Are these vectors already implemented somewhere?

Reply:
Oh, but then it would make testing impossible. There should be some loophole to calculate hashes of byte strings with lengths unknown at compile time